File: Python/Leetcode_Python/Trees/SameTree.py
# Time Complexity: O(n) || Space Complexity: O(n)
from typing import Optional
from Treenode import TreeNode


def isSameTree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    if not p and not q:
        return True
    elif (p and not q) or (q and not p):
        return False
    elif p.val != q.val:
        return False
    else:
        isSameTree(p.left, q.left) and isSameTree(p.right, q.right)

File: Java/Leetcode/Tree/SameTree.java
package com.brianprojects.Leetcode.Tree;

public class SameTree {
    /*
    Given the roots of two binary trees p and q, write a function to check if they are the same or not.
    Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
     */

    public static boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null && q==null){
            return true;
        }

        if(p == null && q != null || p != null && q == null || p.val != q.val){
            return false;
        }

        return (isSameTree(p.left ,q.left) && isSameTree(p.right, q.right));
    }
}
